### ISSUE TITLE
Uses of "do catch" suggest non-working "try" instead

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#53672]
I have code like this
```
    let action : Result<Action, Error> = do catch {
        let signed_action = base64::decode(form.signed_action.as_str())?;
        let signed_action: SignedAction = deserialize_from_slice(signed_action.as_slice())?;
        signed_action.verify(&config.secrets.action_signing_key)?
    };
```
after updating nightly, this does not compile any more. It says
```
error: found removed `do catch` syntax
   --> src/routes.rs:164:42
    |
164 |     let action : Result<Action, Error> = do catch {
    |                                          ^^
    |
    = help: Following RFC #2388, the new non-placeholder syntax is `try`
```

So I tried
```
    let action : Result<Action, Error> = try {
        let signed_action = base64::decode(form.signed_action.as_str())?;
        let signed_action: SignedAction = deserialize_from_slice(signed_action.as_slice())?;
        signed_action.verify(&config.secrets.action_signing_key)?
    };
```
but that does not compile either
```
error: expected identifier, found keyword `let`
   --> src/routes.rs:164:9
    |
163 |     let action : Result<Action, Error> = try {
    |                                          --- while parsing this struct
164 |         let signed_action = base64::decode(form.signed_action.as_str())?;
    |         ^^^ expected identifier, found keyword
```
I have also tried `try catch`, to no avail.

At the very least, the error message about "do catch" needs to be improved, as its advise is clearly incorrect. But also, I wonder what I should do now :D I guess I'll use an immediately-applied closure again...

**Comment by Centril** [comment]
cc @scottmcm

**Comment by scottmcm** [comment]
@RalfJung Oh, right, the error message should mention that you need to be in the 2018 edition.

(I could also make `do catch` or `do try` work in 2015, I suppose...)

**Comment by durka** [comment]
So is `do catch` going to be removed from 2015 with no replacement? That's not very nice.

**Comment by Centril** [comment]
@durka It was never the plan to stabilize `do catch`; it was always a temporary syntax.

**Comment by scottmcm** [comment]
@durka Do you have a case where you have nightly 2015 code that cannot migrate to nightly 2018?

**Comment by jonhoo** [comment]
I'm confused. Even with
```rust
#![feature(rust_2018_preview)]
```
the syntax
```rust
let foo: Result<_, _> = try {
    // ...
};
```
still does not work?

**Comment by jonhoo** [comment]
Ahh, it looks like you *also* need `edition = 2018` in your `Cargo.toml` as outlined in the [transition guide](https://rust-lang-nursery.github.io/edition-guide/editions/transitioning-your-code-to-a-new-edition.html#switch-to-the-next-edition).

**Comment by varkor** [comment]
This was very confusing to run into (for some reason this error occurs on Travis, but not locally for me). It'd make much more sense to me to allow `do catch` before the 2018 edition.

**Comment by scottmcm** [comment]
I've closed this now that there's an additional message about the edition on `try`.  That's not amazing, but this is also a nightly-only thing.

Have you perchance been able to use `try{}` in nightly 2018 edition now, @RalfJung?

**Comment by RalfJung** [comment]
> Have you perchance been able to use `try{}` in nightly 2018 edition now, @RalfJung?

Yes, they seem to work as intended. :)
